### PR TITLE
Add derivation images to parametric map

### PR DESCRIPTION
**Pull request by fedorov**
work towards addressing #186 

At the moment, crashes `itkimage2paramap_makeParametricMap` test

image

Stack trace:

```
#0  0x00007fff86043286 in __pthread_kill ()
#1  0x00007fff860189f9 in pthread_kill ()
#2  0x00007fff8a93d9a3 in abort ()
#3  0x00007fff834271cb in free ()
#4  0x0000000102687cd9 in IODComponent::~IODComponent (this=0x7fff5fbf51f8) at /Users/fedorov/local/builds/dcmqi-new-super/DCMTK/dcmiod/libsrc/modbase.cc:65
#5  0x000000010266612b in DcmIODUtil::freeContainer<OFVector<IODSeriesAndInstanceReferenceMacro::ReferencedSeriesItem*> > (container=@0x7fff5fbf8200) at iodutil.h:840
#6  0x000000010268afa7 in IODCommonInstanceReferenceModule::freeMemory (this=0x7fff5fbf81e0) at /Users/fedorov/local/builds/dcmqi-new-super/DCMTK/dcmiod/libsrc/modcommoninstanceref.cc:353
#7  0x000000010268aea2 in IODCommonInstanceReferenceModule::~IODCommonInstanceReferenceModule (this=0x7fff5fbf81e0) at /Users/fedorov/local/builds/dcmqi-new-super/DCMTK/dcmiod/libsrc/modcommoninstanceref.cc:48
#8  0x000000010268afc5 in IODCommonInstanceReferenceModule::~IODCommonInstanceReferenceModule (this=0x7fff5fbf81e0) at /Users/fedorov/local/builds/dcmqi-new-super/DCMTK/dcmiod/libsrc/modcommoninstanceref.cc:47
#9  0x000000010264d870 in DcmIODCommon::~DcmIODCommon (this=0x7fff5fbf8060) at /Users/fedorov/local/builds/dcmqi-new-super/DCMTK/dcmiod/libsrc/iodcommn.cc:82
#10 0x00000001024fb3f1 in DcmIODImage<IODImagePixelModule<unsigned short>, IODImagePixelModule<short>, IODFloatingPointImagePixelModule, IODDoubleFloatingPointImagePixelModule, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil>::~DcmIODImage (this=0x7fff5fbf8060) at iodimage.h:110
#11 0x00000001024f97bb in DPMParametricMapBase::~DPMParametricMapBase (this=0x7fff5fbf8060) at /Users/fedorov/local/builds/dcmqi-new-super/DCMTK/dcmpmap/libsrc/dpmparametricmapbase.cc:57
#12 0x0000000102505af8 in DPMParametricMapIOD::~DPMParametricMapIOD (this=0x7fff5fbf8060) at /Users/fedorov/local/builds/dcmqi-new-super/DCMTK/dcmpmap/libsrc/dpmparametricmapiod.cc:897
#13 0x0000000102505ba5 in DPMParametricMapIOD::~DPMParametricMapIOD (this=0x7fff5fbf8060) at /Users/fedorov/local/builds/dcmqi-new-super/DCMTK/dcmpmap/libsrc/dpmparametricmapiod.cc:894
#14 0x000000010012466c in OFvariant_destroy_invoker::invoke<DPMParametricMapIOD> (this=0x7fff5fbf38c0, content=0x7fff5fbf8060) at variant.h:197
#15 0x0000000100124598 in OFvariant_invoke_t<OFvariant_destroy_invoker, OFCondition, DPMParametricMapIOD, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil>::operator() (this=0x1001ff038, index=1, content=0x7fff5fbf8060, invoker=@0x7fff5fbf38c0) at variant.h:95
#16 0x00000001001244b0 in OFvariant_invoke<OFvariant_destroy_invoker, OFCondition, DPMParametricMapIOD, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil> (index=1, content=0x7fff5fbf8060, invoker=@0x7fff5fbf38c0) at variant.h:130
#17 0x0000000100124430 in OFvariant<OFCondition, DPMParametricMapIOD, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil>::destroy (this=0x7fff5fbf8060) at variant.h:369
#18 0x00000001001243f5 in OFvariant<OFCondition, DPMParametricMapIOD, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil>::~OFvariant (this=0x7fff5fbf8060) at variant.h:335
#19 0x000000010011ea05 in OFvariant<OFCondition, DPMParametricMapIOD, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil, OFvariadic_nil>::~OFvariant (this=0x7fff5fbf8060) at variant.h:333
#20 0x000000010011a190 in dcmqi::ParaMapConverter::itkimage2paramap (parametricMapImage=@0x7fff5fbf9910, dcmDatasets={<std::__1::__vector_base<DcmDataset *, std::__1::allocator<DcmDataset *> >> = {<std::__1::__vector_base_common<true>> = {<No data fields>}, __begin_ = 0x106729e90, __end_ = 0x106729f30, __end_cap_ = {<std::__1::__libcpp_compressed_pair_imp<DcmDataset **, std::__1::allocator<DcmDataset *>, 2>> = {<std::__1::allocator<DcmDataset *>> = {<No data fields>}, __first_ = 0x106729f30}, <No data fields>}}, <No data fields>}, metaData=@0x7fff5fbf9840) at /Users/fedorov/github/dcmqi/libsrc/ParaMapConverter.cpp:355
#21 0x000000010002f7c7 in ModuleEntryPoint (argc=9, argv=0x7fff5fbff800) at /Users/fedorov/github/dcmqi/apps/paramaps/itkimage2paramap.cxx:59
#22 0x0000000100019bb2 in main (argc=9, argv=0x7fff5fbff800) at /Users/fedorov/local/builds/dcmqi-new-super/SlicerExecutionModel/CMake/SEMCommandLineLibraryWrapper.cxx:33
```

### CODE REVIEW

**Comment by fedorov** [comment]
@michaelonken `addFrame` function is here: https://github.com/QIICR/dcmqi/blob/master/libsrc/ParaMapConverter.cpp#L380

**Comment by michaelonken** [comment]
I don't see anything obvious -- I'll check this tomorrow in detail.

**Comment by fedorov** [comment]
hm, one issue I noticed is plane position FG is added twice, but this is not what I changed - I wonder why this didn't cause any issue in the past: https://github.com/QIICR/dcmqi/blob/master/libsrc/ParaMapConverter.cpp#L432

cc: @che85

**Comment by che85** [comment]
@fedorov Maybe I didn't cause no issues because dcmtk has a mechanism to prevent adding the same FG twice? Just a guess ....

**Comment by fedorov** [comment]
@michaelonken I made some progress. I changed the code to initialize per-frame FGs the same way we do in SEG converter - there is one list of per-frame groups, with a single instance of frame content and image position patient, which are updated for each frame, and derivation image FG inserted/popped out for individual frames as necessary.

I think I need to spend more time debugging the crash. It looks like I am getting a crash independently of whether I add derivation image FG or not (see last commit).

There is some difference in how per-frame FGs were passed in SEG and PM - in SEG we make `new` instances of per-frame FG: https://github.com/QIICR/dcmqi/blob/master/libsrc/ImageSEGConverter.cpp#L329, and in the PM converter (the version before my changes that works!) is using `OFunique_ptr` https://github.com/QIICR/dcmqi/blob/master/libsrc/ParaMapConverter.cpp#L411 - does this make any difference?

As a side note, there is an inconsistency in the API - for segmentation object, we have `addFrame` directly from the segmentation object, and in parametric map it is accessed via `Frames` attribute. Is there a good reason for this?

@che85 moved the code of `addFrame` back into the function to be able easier reproduce the way we populate per-frame FGs in SEG.

Bottom line - this feature is not getting into 1.0.0. Hopefully this will get into a patch sometime soon.

**Comment by fedorov** [comment]
@michaelonken: with the commit above I updated PM converter to (maybe!) make it simpler to debug.

I `#define`d sections of code related to populating CommonInstanceReference and adding derivation FG. When both are disabled (`MAKE_REFS` and `ADD_DERIMG` set to 0 on top of https://github.com/QIICR/dcmqi/pull/193/commits/fd5d8c6c8dc35f6cf8e656ba479096e4b5779075), the test is passing. Setting `MAKE_REFS` or `ADD_DERIMG` to 1 causes a crash.

**Comment by michaelonken** [comment]
Using OFunique_pointer for the functional groups before adding them to the object should not make a difference, it just ensures that the memory of that group is freed once the unique pointer goes out of scope, so you save the delete call at one ore more locations in your code. The dcmfg API copies the functional group for its own use, and (frees its own copy, of course).

From what I see in the stack trace the crash does not happen in functional group code at all but in the Common Instance Reference Module (not impossible something interferes there, though). 

I'll try your code now and let you know what I find out.

**Comment by michaelonken** [comment]
[0001-Fix-crashes-heap-and-double-add-issue.txt](https://github.com/QIICR/dcmqi/files/747774/0001-Fix-crashes-heap-and-double-add-issue.txt)

Ok, here is the result of my analysis:

- refseriesItem must be allocated on the heap since when you add it to the Common Instance Reference Module's Referenced Series, the latter takes over ownership (thus allocating it on the stack will result in a doubled delete, automatic stack deletion and destructor in Common Instance Reference module). That led to a segfault.

- When adding the frames, you create the derivation image FG once, but you add it twice to the per frame functional groups. The same group twice is/should not be a problem when calling addFrame(), however, before going to frame 2 you remove the functional group only once(!) from per frame FG vector and you delete it from memory, however, the dangling pointer of the second addition still stays in the vector. If you then all addFrame() for the second frame, the pointer will be de-referenced inside and you get a segfault.

See the patch file for possible fix (I also moved FGDerivationImage declaration to the rest of the per frame declarations).

Two "issues" remain:
- The Derivation Image Functional Group has to be added for all frames as a per frame functional group, once you add it for a single frame. If for some frames the functional group is empty, that's fine. Actually I wondered that this has gone through dcmfg on write(), I should add/fix a corresponding check.
- For completeness, you should delete the memory of the vector containing the functional groups.

**Comment by fedorov** [comment]
@michaelonken thank you for the patch! It does work for the example I have with per-frame derivation FG populated for each PM frame. Sorry I missed it in my code. I will review SEG converter, because I thought I follow the same approach there.

I attributed the commit from this patch to you to keep the record straight! 👍 

> The Derivation Image Functional Group has to be added for all frames as a per frame functional group, once you add it for a single frame. If for some frames the functional group is empty, that's fine. Actually I wondered that this has gone through dcmfg on write(), I should add/fix a corresponding check.

This is also something I am pretty sure I will need to fix for SEG.

**Comment by michaelonken** [comment]
Thanks :-)

**Comment by fedorov** [comment]
@che85 I think this should be completed - can you take a look?

**Comment by che85** [comment]
Sure, let me check it out and test

File: apps/paramaps/itkimage2paramap.cxx [hunk 1]
```diff
@@ -5,6 +5,13 @@
 #undef HAVE_SSTREAM // Avoid redefinition warning
 #include "dcmqi/ParaMapConverter.h"
 
+#ifdef _WIN32
+#include "dirent_win.h"
+#else
+#include <dirent.h>
+#endif
+
+
```
Comment (che85) [inline apps/paramaps/itkimage2paramap.cxx]: @fedorov maybe we should use the dcmtk provided function for searching directories for files?

Comment (fedorov) [inline apps/paramaps/itkimage2paramap.cxx]: Right, but #191 is still WIP. Depends which one will get merged first.

File: include/dcmqi/ConverterBase.h [hunk 2]
```diff
@@ -260,6 +264,32 @@ namespace dcmqi {
       return 0;
     }
 
+    // AF: I could not quickly figure out how to template this function over image type - suggestions are welcomed!
```
Comment (che85) [inline include/dcmqi/ConverterBase.h]: You can change it to:
```
    template< typename T >
    static vector<vector<int> > getSliceMapForSegmentation2DerivationImage(const vector<DcmDataset*> dcmDatasets,
                                                                           const T &labelImage) {
```

Comment (fedorov) [inline include/dcmqi/ConverterBase.h]: How to refer to `PixelType` inside the function then?

Comment (fedorov) [inline include/dcmqi/ConverterBase.h]: What I really want is to template over `ImageType` as `T`, and then refer to the argument as `T::Pointer`, but this does not compile.

**Comment by fedorov** [comment]
@che85 how about we finish this PR first and then you would rebase and update #191? I see CI is failing for #191, and I am not sure I will have time today to update before I leave.

**Comment by che85** [comment]
Yes let's do this.
